pkg/querier/worker/scheduler_processor.go

```diff
@@ -189,17 +192,61 @@ func (sp *schedulerProcessor) runRequest(ctx context.Context, logger log.Logger,
 		}
 	}
 
-	c, err := sp.frontendPool.GetClientFor(frontendAddress)
-	if err == nil {
-		// Response is empty and uninteresting.
-		_, err = c.(frontendv2pb.FrontendForQuerierClient).QueryResult(ctx, &frontendv2pb.QueryResultRequest{
-			QueryID:      queryID,
-			HttpResponse: response,
-			Stats:        stats,
-		})
-	}
-	if err != nil {
-		level.Error(logger).Log("msg", "error notifying frontend about finished query", "err", err, "frontend", frontendAddress)
+	runPoolWithBackoff(
+		ctx,
+		logger,
+		sp.frontendPool,
+		frontendAddress,
+		func(c client.PoolClient) error {
+			// Response is empty and uninteresting.
+			_, err = c.(frontendv2pb.FrontendForQuerierClient).QueryResult(ctx, &frontendv2pb.QueryResultRequest{
+				QueryID:      queryID,
+				HttpResponse: response,
+				Stats:        stats,
+			})
+			if err != nil {
+				level.Error(logger).Log("msg", "error notifying frontend about finished query", "err", err)
+			}
+			return err
+		},
+	)
+}
+
+var defaultBackoff = backoff.Config{
+	MinBackoff: 100 * time.Millisecond,
+	MaxBackoff: 10 * time.Second,
+	MaxRetries: 5,
+}
+
+func runPoolWithBackoff(
+	ctx context.Context,
+	logger log.Logger,
+	pool *client.Pool,
+	addr string,
+	f func(client.PoolClient) error,
+) {
+	var (
+		backoff = backoff.New(ctx, defaultBackoff)
+		errs    = multierror.New()
+	)
+
+	for backoff.Ongoing() {
+		c, err := pool.GetClientFor(addr)
+		if err != nil {
+			level.Error(logger).Log("msg", "error acquiring client", "err", err)
+			errs.Add(err)
+			pool.RemoveClientFor(addr)
+			backoff.Wait()
+			continue
+		}
+
+		if err = f(c); err != nil {
+			errs.Add(err)
+			pool.RemoveClientFor(addr)
```

Comment:
Would lbe nice to only remove the connection from the pool in a certain error class, but I'm not sure what I'm looking for here.